Adding documentation button for providers new/edit page

alongoldboim opened the pull request.

alongoldboim commented:
This is the general look, although we wont have nothing presented because we don't have documentation yet.
![selection_017](https://cloud.githubusercontent.com/assets/11405684/11894244/92a37efe-a57e-11e5-9081-0863529e2604.png)

simon3z commented:
@miq-bot add_label providers/containers, ui

simon3z commented:
cc @h-kataria @dclarizio

simon3z commented on app/controllers/ems_container_controller.rb:
```diff
@@ -17,4 +17,8 @@ def self.table_name
     redirect_to :action => 'show_list'
   end
+
+  def documantation_url
```
@alongoldboim I assume you used this to test it (and produce the screenshot) but until we won't have a link we may even drop it from the patch.

alongoldboim replied:
@simon3z I agree with you, removing.

simon3z commented on app/helpers/application_helper.rb:
```diff
@@ -7,6 +7,17 @@ module ApplicationHelper
+
+  def documentation_button(url = nil)
+    hidden_div_if(url.nil?, :class => "documentation") do
+      content_tag(:span, "* For more inforemation   ") +
```
@alongoldboim typo "information"

Fryguy commented on app/controllers/ems_common.rb:
```diff
@@ -258,6 +258,7 @@ def show_list
   end
 
   def new
+    @doc_url = self.try(:documantation_url)
```
Typo...should be `documentation_url`

Fryguy replied:
Also, the try is not necessary...there's no way self can be nil.

alongoldboim replied:
@Fryguy your right, changed.

Fryguy commented on app/helpers/application_helper.rb:
```diff
@@ -7,6 +7,17 @@ module ApplicationHelper
   include StiRoutingHelper
   include TextualSummaryHelper
 
+  def icon_text(i, text = "", opts = {})
```
I feel like this method already exists somewhere, but I'm not sure.  @epwinchell @h-kataria ?

Fryguy replied:
@himdel ?

himdel replied:
Uh, I think hardcoding `glyphicon glyphicon-` may be problematic, since we're using 4 different icon fonts - fontawesome, glyphicon, pficon and product (whatever that one is).

That said, there is a `listicon_glyphicon_tag` in [application_helper](https://github.com/ManageIQ/manageiq/blob/master/app/helpers/application_helper.rb#L1339), which tries to do a similar thing but the bulk of it's code is a big case of cases.

There's also `toolbar_image` in [toolbar_helper](https://github.com/ManageIQ/manageiq/blob/master/app/helpers/toolbar_helper.rb#L105) which is probably too specific.

Aand finally `active_icon` / `inactive_icon` in [configuraion_view_helper](https://github.com/ManageIQ/manageiq/blob/master/app/helpers/configuration_helper/configuration_view_helper.rb#L22), equally as specific.

Some I'm OK with such function, but I'd prefer more flexibility in the class name, so that non-glyphicon icons can be used as well.

himdel replied:
Something like `icon_text(kind, i, ...` with the class name being generated as `"#{kind} #{kind}-#{i}"` would work. (Even though `.product` and `.pficon` don't exist by themselves.)

moolitayer commented on app/helpers/application_helper.rb:
```diff
@@ -7,6 +7,17 @@ module ApplicationHelper
+  end
+
+  def documentation_button(url = nil)
+    hidden_div_if(url.nil?, :class => "documentation") do
```
do we need a div at all (hidden) when there is no url?

alongoldboim replied:
@moolitayer Not really when thinking about it :).
Changed.

simon3z commented:
@h-kataria @dclarizio any comment about the UI?

@moolitayer @zeari LGTY?

Fryguy commented:
@himdel Can you also review?

himdel commented on app/controllers/ems_common.rb:
```diff
@@ -258,6 +258,7 @@ def show_list
   end
 
   def new
+    @doc_url = self.documentation_url if respond_to? :documentation_url
```
This should probably happen in `edit` as well as `new`. And possibly `create`, though I'm not sure about that one..

himdel commented:
LGTM, assuming a more generic `icon_text` so that we can use that in other places as well.. 

..and assuming it works with edit as well as new, which it currently doesn't.

dclarizio commented:
I like the idea and I think we may be able to find use of this in other places, but where will this button point to?  A website (Red Hat?) that holds the doc or maybe a wiki page that describes what the token is?  Would a "?" image with pop-up text work just as well?

>  in the current ui the button will not be visible.

Why are we adding a button that won't be visible?

@epwinchell Is there a patternfly button style we should be using here?

alongoldboim commented:
@himdel @simon3z Good comments, changed the icon_text to be more generic (we will be able to use it with other icons then glyphicon), added documentation also to edit.
@dclarizio Im sure we can use it in other places to simplify things, regarding where it should point if we don't have documentation on how to get the token from various providers maybe we should add a doc for it?

simon3z commented:
LGTM :+1: cc @dclarizio @h-kataria @chessbyte 

@dclarizio what is missing to merge this one?

epwinchell commented:
@serenamarie125 Can you give your thoughts on how we might link to documentation within the Product? I think a simple link reading "See Documentation" would look cleaner than a button.

alongoldboim commented:
@matthewd @dclarizio @himdel @chessbyte 
Added a general documentation link regarding provider, this documentation btn will be added to all new/edit provider forms.
![selection_044](https://cloud.githubusercontent.com/assets/11405684/12111844/3370e08a-b3a1-11e5-992f-a4cb3f0815d8.png)

simon3z commented:
@serenamarie125 could you review the screenshots? Thanks!

abonas commented:
@alongoldboim this PR adds a general capability documentation, but not for the container provider token one (at least I didn't spot it in the commit). If that's correct, I suggest changing the commit's and the PRs title to reflect more accurately what it does.
Also, will this remain in downstream, navigating to the manageiq.org?

@chessbyte, while at it, the link in this commit leads to a page that is very infra centered ("Navigate to Infrastructure → Providers.") - which is not a good fit when linked from cloud and containers providers. Any plans to update the page to a more generic one?

alongoldboim commented:
@abonas 
* Changed the desc of this pr.
* It will be changed in downstream.
* Changed the text.

abonas commented on app/helpers/application_helper.rb:
```diff
@@ -7,6 +7,21 @@ module ApplicationHelper
+  def documentation_button(url = nil)
+    if url
+      content_tag(:div, :class => "documentation") do
+        content_tag(:span, "* For more information visit/read the  ") +
```
does it need a translation key/entry?

abonas commented:
> Changed the text.

great. consider an updated screenshot

dclarizio commented:
@alongoldboim I'm pinging @serenamarie125 once more as I'd like to discuss if we want to go down this path of having documentation buttons/links sprinkled throughout the UI or if there is a more standard way of introducing inline help.

serenamarie125 commented:
This should just be a link - rather than text + a button ... 
The link should follow the PatternFly visuals & behavior and should say 

For more information, visit theLinkName

* if you need help with PatternFly visuals, let me know.  Should be blue & upon hover, blue should change and entire string should be underlined

alongoldboim commented:
@simon3z @serenamarie125 @dclarizio changed, please recheck.
![doqwcn](https://cloud.githubusercontent.com/assets/11405684/12146204/0f9bd6ec-b49b-11e5-8c40-486ac050c13e.gif)

epwinchell commented:
@alongoldboim I think only the word "Documentation" should be linked

alongoldboim commented:
@epwinchell done.

himdel commented:
@alongoldboim uh, pretty sure that the favicon should be before "For more info.." and not in the middle of the text.

epwinchell commented:
@alongoldboim Check with @serenamarie125 but, I think the idea was to use text only, not an icon.

alongoldboim commented:
@serenamarie125 
* Should i do it without the icon?
* In case we want to use the icon where should it be placed?

serenamarie125 commented:
It should just be a single link, including all text "For more information, visit the Containers Providers documentation."
And do not include an icon.
This is something that we will be introducing in PatternFly going forward.  Thanks!

alongoldboim commented:
Done.
![selection_001](https://cloud.githubusercontent.com/assets/11405684/12215524/c8214790-b6c8-11e5-806f-68f6b685e5a0.png)

chessbyte commented on app/controllers/ems_common.rb:
```diff
@@ -314,6 +315,7 @@ def create
   end
 
   def edit
+    @doc_url = provider_documentation_url if respond_to? :provider_documentation_url
```
When would the `if` ever be `false`?  If never, why have the `if` at all?

alongoldboim replied:
Your right, removed.

chessbyte commented on app/controllers/ems_common.rb:
```diff
@@ -561,6 +563,10 @@ def button
   end
 
+  def provider_documentation_url
+    "http://manageiq.org/documentation/getting-started/#adding-a-provider"
```
I would have thought that the `url` would be maintained by each EMS model.  Not sure I understand what having it hard-coded here does.

alongoldboim replied:
Each EMS model can maintain its own url, defining the same method inside a specific provider will override this method, this url is for a generic provider.

chessbyte replied:
The same-named method in an EMS model would *not* override this method, because this one is being the *controller*, not the *model*.

alongoldboim replied:
meant EMS controller*. (e.g defining it again in EmsContainerController will take effect over the container provider new/edit documentation link)

Fryguy commented:
Removing the containers label because this is not a containers-only PR.
